runtime/runc_jailer.go

```diff
@@ -0,0 +1,370 @@
+// Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"syscall"
+
+	"github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/firecracker-microvm/firecracker-go-sdk"
+
+	"github.com/firecracker-microvm/firecracker-containerd/internal/vm"
+)
+
+// runcJailer uses runc to set up a jailed environment for the Firecracker VM.
+type runcJailer struct {
+	logger         *logrus.Entry
+	ociBundlePath  string
+	runcBinaryPath string
+	uid            uint32
+	gid            uint32
+}
+
+func newRuncJailer(logger *logrus.Entry, ociBundlePath, runcBinPath string, uid, gid uint32) *runcJailer {
+	return &runcJailer{
+		logger:         logger,
+		ociBundlePath:  ociBundlePath,
+		runcBinaryPath: runcBinPath,
+		uid:            uid,
+		gid:            gid,
+	}
+}
+
+// JailPath returns the base directory from where the jail binary will be ran
+// from
+func (j runcJailer) OCIBundlePath() string {
+	return j.ociBundlePath
+}
+
+// RootPath returns the root fs of the jailed system.
+func (j runcJailer) RootPath() string {
+	return filepath.Join(j.OCIBundlePath(), rootfsFolder)
+}
+
+func (j runcJailer) JailPath() vm.Dir {
+	return vm.Dir(j.RootPath())
+}
+
+func (j *runcJailer) BuildJailedMachine(cfg *Config, machineConfig *firecracker.Config, vmID string) ([]firecracker.Opt, error) {
+	handler := j.BuildHandler(cfg, &machineConfig.SocketPath, vmID)
+	fifoHandler := j.BuildLinkFifoHandler()
+	client := firecracker.NewClient(machineConfig.SocketPath, j.logger, machineConfig.Debug)
+
+	opts := []firecracker.Opt{
+		firecracker.WithProcessRunner(j.jailerCommand(vmID)),
+		firecracker.WithClient(client),
+		func(m *firecracker.Machine) {
+			m.Handlers.FcInit = m.Handlers.FcInit.Prepend(handler)
+			m.Handlers.FcInit = m.Handlers.FcInit.AppendAfter(firecracker.CreateLogFilesHandlerName, fifoHandler)
```

Comment:
Why after that one specifically?  This seems brittle as a change elsewhere (removing that handler or calling `AppendAfter` that handler again) can change the ordering.

Reply:
That's due to linking of the fifos to where firecracker can see it. We can only link the files after they exist, meaning after they've been created which is the `CreateLogFileHandlerName`. The ordering shouldn't matter, since the fifos need to be created before the machine has started. So this can be the first handler or last, it shouldn't make a difference.

Reply:
Please write a comment in the code.

Reply:
@xibz Can you address this comment?